### PR TITLE
Post on system calls etc

### PR DESCRIPTION
**Pull request by maelle**

### CODE REVIEW

File: content/blog/2021-07-26-system-calls-r-package/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+## Using the tool without calling the system
+
+Before jumping to find ways to control that needed tool with system commands, you might try and find ways to not have to do that.
+
+* Maybe there's a C or C++ library offering similar features to the CLI, like ImageMagick++, wrapped in the [magick R package](https://docs.ropensci.org/magick/) does for [ImageMagick](https://imagemagick.org/index.php); or like [libtiff wrapped in the ijtiff R package](/blog/2018/04/12/ijtiff/)? To get started with wrapping C libraries, check out [Davis Vaughan's blog post](https://blog.davisvaughan.com/2019/03/02/now-you-c-me/). For C++ dive into either [Rcpp](https://github.com/RcppCore/Rcpp) or the more recent [cpp11](https://cpp11.r-lib.org/articles/cpp11.html) (see [a vignette of cpp11](https://cpp11.r-lib.org/articles/motivations.html) for the motivations of cpp11 in particular compared to Rcpp).
+* In the case of a Python module, you can use the [reticulate package](https://rstudio.github.io/reticulate/) for interacting with Python.
+* In the case of a JS library, you can use the [V8 package](https://github.com/jeroen/v8).
```
Comment (maelle) [inline content/blog/2021-07-26-system-calls-r-package/index.Rmd:29]: I am wondering whether to add something about Java. rJava being difficult to install, an approach like nlrx https://github.com/ropensci/nlrx/ that runs a Java virtual machine without rJava is interesting.

File: content/blog/2021-07-26-system-calls-r-package/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,77 @@
+---
+title: How to Use System Commands in your R Package
```
Comment (maelle) [inline content/blog/2021-07-26-system-calls-r-package/index.Rmd:2]: Maybe it shouldn't even have the word "package".

File: content/blog/2021-07-26-system-calls-r-package/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+author:
+  - Maëlle Salmon
+  - Jeroen Ooms
+date: '2021-07-26'
```
Comment (maelle) [inline content/blog/2021-07-26-system-calls-r-package/index.Rmd:6]: I want to publish this _after_ Friday newsletter to ensure some content for the next newsletter :sweat_smile:

File: content/blog/2021-07-26-system-calls-r-package/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+slug: system-calls-R-package
+categories: []
+tags:
```
Comment (maelle) [inline content/blog/2021-07-26-system-calls-r-package/index.Rmd:9]: I need to add the tech notes tag.

**Comment by jeroen** [comment]
## Shelling out or not

Interfacing to a C/C++ API is usually more faster and more robust than executing system commands.

The core issue is that invoking a command always starts a new process, whereas a C/C++ library is loaded in the current process. Different processes cannot share memory, hence all interaction has to happen via text streams.

### Input

A C/C++ API can have many functions with one or more arguments consisting of typed data structures. 

On the other hand, the only way to pass input to a shell command is a long string that contains all the command line arguments. There is no way to share memory between R and the program we are calling. To pass any data to the program, you probably need to write it to disk in some format, and then the program has to read it from the file, which can lead to performance overhead or read/write errors. 

When the user has can provide the arguments (eg filenames) that are passed to the subprogram, you have to consider these names may contain spaces or non-ascii characters. This can again lead to problems when they are all joined in a single string and executed.

### Output and errors

The output of a command line program only consists of a status code (indicating if the program had an error or not) and a stream of text that the program prints to the screen. 

There is no return value, or exception handling if an error appears in program that we called. This means that if we wrap a command line program in R, we have very limited means to handle errors. Basically the only thing we can in case when the program returns an error, is show the text that the program printed to the user.

**Comment by mpadge** [comment]
@jeroen Just a though, but do you want to add a warning about the strongly increased dangers of making external calls within parallel processes (or code every likely to be used in such)? Or a single line strongly advising against doing that?

**Comment by jeroen** [comment]
@mpadge I don't think that has to be a problem per se? What sort of dangers do you mean?

**Comment by mpadge** [comment]
@jeroen In trying to get `pkgstats` to trawl CRAN in parallel, I tried every parallel-ing/multicore framework that exists for R, but the external `ctags`/`gtags` processes that are called without supervision get killed frequently enough that it corrupts their (file) output and makes them unusable. It seems this happens because [`mc_cleanup` in the parallel library](https://github.com/wch/r-source/blob/78422c2487633c9ce7b0f29b5f0b4859c96314c7/src/library/parallel/src/fork.c#L364) ends up [killing any detached child processes](https://github.com/wch/r-source/blob/78422c2487633c9ce7b0f29b5f0b4859c96314c7/src/library/parallel/src/fork.c#L404).

The [`parallel/src` code explains how this can be avoided](https://github.com/wch/r-source/blob/78422c2487633c9ce7b0f29b5f0b4859c96314c7/src/library/parallel/src/fork.c#L334-L361), but from my observations it would seem that no parallel wrappers up to and including `future` actually expose an ability to set an `sKill` flag. (I could be wrong there, and haven't looked in the actual code of `future` or the likes - maybe it is possible, but definitely not in any "default mode".) Anyway, without explicit ability to set an `sKill` flag, detached processes will always be killed at what is effectively an arbitrary time, and since any parallel jobs calling such processes by definition must need some kind of result from them, that result can never be presumed generally accessible. That's my current conclusion anyway.

**Comment by maelle** [comment]
@mpadge should this challenge be its own tech note?

**Comment by jeroen** [comment]
Yeah this might be out of scope for this technote. I'd be interested to see if the problem also appears if you use sys instead of system2.

Why are you using parallel in the first place? If you need to make many concurrent calls to gtags, maybe you can better use processx do make all those calls?

**Comment by mpadge** [comment]
Yeah, it's definitely out of scope for this note, but a mere warning line about potential problems of `system`/`sys` in parallel processes might provide a useful reference as a warning. The problems I encounter came from various parallel libs applied to standard `system` calls. `pkgstats` doesn't use `processx`, because it's also a core part of `pkgcheck` which does, as does the API to deliver new bot check results, and embedding processes triggered from separate packages is also potentially problematic. Because of this embedding, `pkgstats` has to act as much as possible like a simple single-process call, even though it makes these 2 external calls. Then again, they really are special cases because the main libraries are not intended to be directly called like that anyway, and so provide no way of connecting a supervising process. `processx` could also do nothing with them, because it would return straight away with a term signal. Anyway, just maybe a word of two of warning just in case was about all i was thinking.

**Comment by jeroen** [comment]
Hmm I'd like to debug this a bit more before warning against it because it sounds to me R should only kill the programs if somehow they are still alive after the R itself has quit. That shouldn't happen in your case, right?

When you have time, could you maybe test if swapping out system for sys makes any difference?

**Pull request merged**